### PR TITLE
feat(map): cluster&rotate bus

### PR DESCRIPTION
**Pull request by NoamGaash**
image

### CODE REVIEW

**Review by ReallyLiri — COMMENTED** [review]
remove package-lock.json (or yarn.lock), we don't want both.

File: src/api/agencyList.ts [hunk 1]
```diff
@@ -0,0 +1,11 @@
+export interface Agency {
+  date: string // example - "2019-07-01"
+  operator_ref: number // example - 25,
+  agency_name: string // example - "אלקטרה אפיקים"
```
Comment (ReallyLiri) [inline src/api/agencyList.ts:4]: we already have a model for this - https://github.com/hasadna/open-bus-map-search/blob/main/src/model/operator.ts
let's merge them.

File: src/api/useVehicleLocations.ts [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+ * this is a custom hook that fetches the vehicle locations from the API.
+ * it recieves an interval of two dates, and loads locations of all vehicles in that interval.
+ * if some of the interval has already been loaded,
+ */
+
+import { useEffect, useState } from 'react'
+import { VehicleLocation } from 'src/model/vehicleLocation'
+
+const config = {
+  apiUrl: 'https://open-bus-stride-api.hasadna.org.il/siri_vehicle_locations/list?get_count=false',
```
Comment (ReallyLiri) [inline src/api/useVehicleLocations.ts:11]: let's use the generated api? we don't want to maintain a couple of api access code

File: src/api/useVehicleLocations.ts [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+  limit: 500, // the maximum number of vehicles to load in one request
```
Comment (ReallyLiri) [inline src/api/useVehicleLocations.ts:12]: move it to a const with meaningful name instead of a comment.

File: src/api/useVehicleLocations.ts [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+  const observable = loadedLocations.get(key)!
+  return observable.observe(update)
+}
+
+function getMinutesInRange(from: Dateable, to: Dateable) {
+  const start = new Date(from).setSeconds(0, 0)
+  const end = new Date(to).setSeconds(0, 0)
+
+  // array of minutes to load
+  const minutes = Array.from({ length: (end - start) / 60000 }, (_, i) => ({
```
Comment (ReallyLiri) [inline src/api/useVehicleLocations.ts:100]: nothing in Moment.js or similar to get this logic?

File: src/api/useVehicleLocations.ts [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+export default function useVehicleLocations({ from, to }: { from: Dateable; to: Dateable }) {
+  const [locations, setLocations] = useState<VehicleLocation[]>([])
+  useEffect(() => {
+    const unmounts = getMinutesInRange(from, to).map(({ from, to }) =>
+      getLocations(from, to, (locations) => {
+        setLocations(
+          (prev) =>
+            [...prev, ...locations]
+              .sort((a, b) => a.id - b.id)
+              .filter((loc, i, arr) => arr.findIndex((l) => l.id === loc.id) === i), // remove duplicates
```
Comment (ReallyLiri) [inline src/api/useVehicleLocations.ts:116]: can use lodash

File: src/model/vehicleLocation.ts [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+  siri_ride__scheduled_start_time: string
+  siri_ride__vehicle_ref: string
+  siri_ride__first_vehicle_location_id: number
+  siri_ride__last_vehicle_location_id: number
+  siri_ride__duration_minutes: number
+  siri_ride__gtfs_ride_id: number
+  siri_route__id: number
+  siri_route__line_ref: number
+  siri_route__operator_ref: number
+}
```
Comment (ReallyLiri) [inline src/model/vehicleLocation.ts:37]: use generated model instead of doing it yourself

File: src/pages/RealtimeMapPage.tsx [hunk 7]
```diff
@@ -60,66 +65,87 @@ function numberToColorHsl(i: number, max: number) {
...
+                setFrom(e.target.value)
+                setTo(formatTime(+new Date(e.target.value) + (+new Date(to) - +new Date(from))))
+              }}
+            />
           </label>{' '}
           {` `}
           <label>
-            עד תאריך
-            <input type="datetime-local" value={to} onChange={(e) => setTo(e.target.value)} />
+            צפה במיקומי אוטובוסים בטווח של
```
Comment (ReallyLiri) [inline src/pages/RealtimeMapPage.tsx:121]: I tried to keep all strings in one place (texts.ts or something) to make translation and compatibility easier.

File: src/pages/RealtimeMapPage.tsx [hunk 8]
```diff
@@ -60,66 +65,87 @@ function numberToColorHsl(i: number, max: number) {
...
+              onChange={(e) => setTo(formatTime(+new Date(from) + +e.target.value * 1000 * 60))}
+            />
+            דקות
           </label>
         </div>
+        <div className="map-header-buttons">
+          <button
+            onClick={() => {
+              setFrom(formatTime(+new Date() - 5 * 1000 * 60))
+              setTo(formatTime(+new Date() - 4 * 1000 * 60))
```
Comment (ReallyLiri) [inline src/pages/RealtimeMapPage.tsx:134]: would be really nice to use some helper methods or better yet a library for these converts, they're not readable

**Comment by ReallyLiri** [comment]
I published new version of generated client: https://www.npmjs.com/package/open-bus-stride-client v0.1.3
it should contain any new models and apis.

**Comment by NoamGaash** [comment]
@ReallyLiri thaks for the review! Great comments, I'll fix it tomorrow in the workshop 🥳

File: src/api/agencyList.ts (hunk 1 above)
Comment (NoamGaash) [reply src/api/agencyList.ts:4]: done

File: src/api/useVehicleLocations.ts (hunk 2 above)
Comment (NoamGaash) [reply src/api/useVehicleLocations.ts:11]: can't find such thing

File: src/api/useVehicleLocations.ts (hunk 4 above)
Comment (NoamGaash) [reply src/api/useVehicleLocations.ts:100]: momentJS is deprecated, I try to use it as little as possible

File: src/model/vehicleLocation.ts (hunk 6 above)
Comment (NoamGaash) [reply src/model/vehicleLocation.ts:37]: I don't know how to

File: src/pages/RealtimeMapPage.tsx (hunk 7 above)
Comment (NoamGaash) [reply src/pages/RealtimeMapPage.tsx]: done

File: src/pages/RealtimeMapPage.tsx (hunk 8 above)
Comment (NoamGaash) [reply src/pages/RealtimeMapPage.tsx:135]: I don't feel it's not readable - I'll add a comment anyway

**Pull request merged**